src/libraries/System.Private.CoreLib/src/System/IO/Strategies/FileStreamHelpers.cs

```diff
@@ -126,9 +120,29 @@ internal static void ValidateArguments(string path, FileMode mode, FileAccess ac
                 throw new ArgumentException(SR.Argument_InvalidAppendMode, nameof(access));
             }
 
+            if (preallocationSize > 0)
+            {
+                ValidateArgumentsForPreallocation(mode, access);
+            }
+
             SerializationGuard(access);
         }
 
+        internal static void ValidateArgumentsForPreallocation(FileMode mode, FileAccess access)
+        {
+            if ((access & FileAccess.Write) == 0)
+            {
+                throw new ArgumentException(SR.Argument_InvalidPreallocateAccess, nameof(access));
+            }
```

Comment:
I think throwing opposes to the whole point of treating it as a hint and we have discussed it in the past https://github.com/dotnet/runtime/pull/58726#discussion_r705292696.

Reply:
Treating it as a hint is about ignoring when the platform does not support it.

Throwing ArgumentException is about helping the user understand how to use the API.